packages/vite/src/node/plugins/asset.ts

```diff
@@ -222,10 +222,31 @@ async function fileToBuiltUrl(
     if (!map.has(contentHash)) {
       const basename = path.basename(file)
       const ext = path.extname(basename)
-      const fileName = path.posix.join(
+      let fileName = path.posix.join(
         config.build.assetsDir,
         `${basename.slice(0, -ext.length)}.${contentHash}${ext}`
       )
+
+      const { output } = config.build?.rollupOptions ?? {}
+      // Only the object format is currently considered here.
+      if (output && !Array.isArray(output)) {
+        const assetFileNames = output.assetFileNames
+        // e.g assetFileNames: dir/[name].[ext]
+        if (typeof assetFileNames === 'string') {
+          fileName = assetFileNames
+            .replace(/\[name\]/, basename.slice(0, -ext.length))
+            .replace(/\[ext\]/, ext.trimLeft('.'))
+            .replace(/\[hash\]/, getAssetHash(content))
```

Comment:
Check out https://rollupjs.org/guide/en/#outputassetfilenames, looks like the user can also specify `[extname]`

> [extname]: The file extension of the asset including a leading dot, e.g. .css.
> [ext]: The file extension without a leading dot, e.g. css.
> [hash]: A hash based on the name and content of the asset.
> [name]: The file name of the asset excluding any extension.